implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,237 @@
+ *
+ * <ol>
+ * <li>file or directory</li>
+ * <li>classpath resource</li>
+ * <li>jar resource</li>
+ * <li>http resource</li>
+ * </ol>
+ * <p>
+ *
+ * If a profile is active, the profile resource is loaded if available and the unprofiled resource is also
+ * available. This is to keep a consistent order with the unprofiled resource. Profiles are not taken into account if
+ * the location is a directory.
+ */
+@Experimental("Load additional config locations")
+public abstract class AbstractLocationConfigSourceFactory implements ConfigSourceFactory {
+    public static final String SMALLRYE_LOCATIONS = "smallrye.config.locations";
+
+    protected abstract String[] getFileExtensions();
+
+    protected abstract ConfigSource loadConfigSource(final URL url, final int ordinal) throws IOException;
+
+    protected ConfigSource loadConfigSource(final URL url) throws IOException {
+        return this.loadConfigSource(url, ConfigSource.DEFAULT_ORDINAL);
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context) {
+        final ConfigValue value = context.getValue(SMALLRYE_LOCATIONS);
+
+        final List<ConfigSource> configSources = new ArrayList<>();
+        if (value.getValue() != null) {
+            List<URL> resources = newCollectionConverter(new URLConverter(), ArrayList::new).convert(value.getValue());
+            if (resources != null) {
+                for (URL url : resources) {
+                    tryFileSystem(url, configSources, context);
+                    tryClassPath(url, configSources, context);
+                    tryHttpResource(url, configSources, context);
+                }
+            }
+        }
+
+        return configSources;
+    }
+
+    private void tryFileSystem(final URL url, final List<ConfigSource> configSources, final ConfigSourceContext context) {
+        if (url.getProtocol().startsWith("file")) {
+            // Load single file and profiles
+            final Path urlPath = Paths.get(url.getFile()).toAbsolutePath().normalize();
+            if (Files.isRegularFile(urlPath)) {
+                consumeAsPath(toURL(urlPath.toUri()), new ConfigSourcePathConsumer(this, configSources, context));
+            } else if (Files.isDirectory(urlPath)) { // Load directory contents
+                try (Stream<Path> paths = Files.walk(urlPath, 1)) {
+                    paths
+                            .filter(Files::isRegularFile)
+                            .filter(path -> validExtension(path.getFileName().toString()))
+                            .forEach(path -> addConfigSource(toURL(path.toUri()), configSources));
+                } catch (IOException e) {
+                    throw ConfigMessages.msg.failedToLoadResource(e);
+                }
+            }
+        }
+    }
+
+    private void tryClassPath(final URL url, final List<ConfigSource> configSources, final ConfigSourceContext context) {
+        if (url.getProtocol().startsWith("file")) {
```

Comment:
Using relative URIs, you can check to see if the URI has a `null` scheme _or_ the scheme is equal to `file`; in the former case, it's a relative URI.